theano/tensor/nnet/tests/test_abstract_conv.py

```diff
@@ -406,3 +410,171 @@ def test_grad_types(self):
             grad_filters, grad_filters.type, filters, filters.type)
         assert grad_topgrad.type == topgrad.type, (
             grad_topgrad, grad_topgrad.type, topgrad, topgrad.type)
+
+
+class TestBilinearUpsampling(unittest.TestCase):
+
+    def numerical_kernel_1D(self, ratio):
+        """Gets numerical 1D kernel for bilinear upsampling"""
+        return np.array([i for i in range(1, ratio + 1) +
+                         range(ratio - 1, 0, -1)])
+
+    def numerical_kernel_2D(self, ratio):
+        """Gets numerical 2D kernel for bilinear upsampling"""
+        return np.array([i * j for i in self.numerical_kernel_1D(ratio) for j
+                         in self.numerical_kernel_1D(ratio)]).\
+            reshape(2 * ratio - 1, 2 * ratio - 1)
+
+    def test_bilinear_kernel_2D(self):
+        """Test 2D kernels used in bilinear upsampling
+
+        This method tests the correctness of the
+        2D kernel values used in bilinear upsampling
+        for some upsampling ratios.
+
+        """
+        for ratio in [2, 3, 4, 5, 6, 7, 8, 9]:
+            # getting the un-normalized kernel
+            kernel = bilinear_kernel_2D(ratio=ratio, normalize=False)
+            f = theano.function([], kernel)
+            kernel_2D = self.numerical_kernel_2D(ratio)
+            np.testing.assert_allclose(kernel_2D, f())
+
+            # getting the normalized kernel
+            kernel = bilinear_kernel_2D(ratio=ratio, normalize=True)
+            f = theano.function([], kernel)
+            kernel_2D = kernel_2D / float(ratio**2)
+            np.testing.assert_allclose(kernel_2D, f())
+
+    def test_bilinear_kernel_1D(self):
+        """Test 1D kernels used in bilinear upsampling
+
+        This method tests the correctness of the
+        1D kernel values used in bilinear upsampling
+        for some upsampling ratios.
+
+        """
+        for ratio in [2, 3, 4, 5, 6, 7, 8, 9]:
```

Comment:
add `ratio = as_tensor_variable(ratio)`, so that we check that it works for tensors